src/py21cmfast/drivers/single_field.py

```diff
@@ -0,0 +1,1268 @@
+            halobox,
+        ),
+        astro_params=astro_params,
+        flag_options=flag_options,
+    )
+
+    if spin_temp is not None and not inputs.flag_options.USE_TS_FLUCT:
+        logger.warning(
+            "Changing flag_options.USE_TS_FLUCT to True since spin_temp was passed."
+        )
+        flag_options.USE_TS_FLUCT = True
+
+    # Get the previous redshift
+    if previous_ionized_box is not None:
+        prev_z = previous_ionized_box.redshift
+
+        # Ensure the previous ionized box has a higher redshift than this one.
+        if prev_z <= redshift:
+            raise ValueError(
+                "Previous ionized box must have a higher redshift than that being evaluated."
+            )
+    elif (
+        inputs.flag_options.INHOMO_RECO
+        or inputs.flag_options.USE_TS_FLUCT
+        and redshift >= global_params.Z_HEAT_MAX
+    ):
+        prev_z = 0  # signal value for first box
+    else:
+        raise ValueError(
+            "You need to provide a previous ionized box when redshift < Z_HEAT_MAX."
+        )
+
+    box = IonizedBox(
+        redshift=redshift,
+        random_seed=random_seed,
+        inputs=inputs,
+        prev_ionize_redshift=prev_z,
+    )
+
+    # Construct FFTW wisdoms. Only if required
+    construct_fftw_wisdoms(
+        user_params=inputs.user_params, cosmo_params=inputs.cosmo_params
+    )
+
+    # Check whether the boxes already exist
+    if not regenerate:
+        with contextlib.suppress(OSError):
+            box.read(direc)
+            logger.info(
+                f"Existing z={redshift} ionized boxes found and read in (seed={box.random_seed})."
+            )
+            return box
+
+    # EVERYTHING PAST THIS POINT ONLY HAPPENS IF THE BOX DOESN'T ALREADY EXIST
+    # ------------------------------------------------------------------------
+
+    # Get appropriate previous ionization box
+    if previous_ionized_box is None:
+        previous_ionized_box = IonizedBox(
+            redshift=0, flag_options=inputs.flag_options, initial=True
+        )
+
+    if previous_perturbed_field is None:
+        # If we are beyond Z_HEAT_MAX, just make an empty box
+        if prev_z == 0:
+            previous_perturbed_field = PerturbedField(
+                redshift=0, user_params=inputs.user_params, initial=True
+            )
+        else:
+            raise ValueError("No previous perturbed field given, but one is required.")
+
+    if not flag_options.USE_HALO_FIELD:
+        # Construct an empty halo field to pass in to the function.
+        halobox = HaloBox(
+            redshift=0,
+            inputs=inputs,
+            dummy=True,
+        )
+    elif halobox is None:
+        raise ValueError("No halo box given but USE_HALO_FIELD=True")
+
+    # Set empty spin temp box if necessary.
+    if not flag_options.USE_TS_FLUCT:
+        spin_temp = TsBox(
+            redshift=0,
+            inputs=inputs,
+            dummy=True,
+        )
```

Comment:
Hmmm, maybe that's a good idea... I'm struggling to think of the pros vs. cons. How would you implement it?

Reply:
I guess it would be a part of `get_required_input_arrays` where structures which are arguments of the backend but not required return as dummy structs

Reply:
Noted for another PR